src/main/java/com/intuit/graphql/orchestrator/batch/NoExternalReferenceSelectionSetModifier.java

```diff
@@ -2,28 +2,40 @@
 
 import static com.intuit.graphql.orchestrator.resolverdirective.FieldResolverDirectiveUtil.hasResolverDirective;
 import static graphql.introspection.Introspection.TypeNameMetaFieldDef;
+import static graphql.schema.FieldCoordinates.coordinates;
+import static graphql.util.TreeTransformerUtil.changeNode;
 import static graphql.util.TreeTransformerUtil.deleteNode;
 import static java.util.Objects.requireNonNull;
 
+import com.intuit.graphql.orchestrator.federation.metadata.FederationMetadata;
+import com.intuit.graphql.orchestrator.federation.metadata.FederationMetadata.EntityMetadata;
+import com.intuit.graphql.orchestrator.schema.ServiceMetadata;
 import graphql.language.Field;
 import graphql.language.FragmentDefinition;
 import graphql.language.InlineFragment;
 import graphql.language.Node;
 import graphql.language.NodeVisitorStub;
 import graphql.language.SelectionSet;
+import graphql.schema.FieldCoordinates;
 import graphql.schema.GraphQLFieldDefinition;
 import graphql.schema.GraphQLFieldsContainer;
 import graphql.schema.GraphQLType;
 import graphql.schema.GraphQLTypeUtil;
 import graphql.util.TraversalControl;
 import graphql.util.TraverserContext;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.commons.collections4.CollectionUtils;
 
 public class NoExternalReferenceSelectionSetModifier extends NodeVisitorStub {
 
-  private GraphQLFieldsContainer rootType;
+  private final GraphQLFieldsContainer rootType;
+  private final ServiceMetadata serviceMetadata;
 
-  NoExternalReferenceSelectionSetModifier(GraphQLFieldsContainer rootType) {
+  NoExternalReferenceSelectionSetModifier(GraphQLFieldsContainer rootType, ServiceMetadata serviceMetadata) {
```

Comment:
Renaming this class?

Reply:
yeah we should rename this class.  This class is now not just an external field remover since we add fields to it
1.  NodeVisitorSelectionSetModifier
2.  DownstreamQueryModifier

Reply:
was renamed to DownstreamQueryModifier